File: tbsrc-golang/pkg/types/threshold_set.go
package types

// ThresholdSet 对应 C++ struct ThresholdSet
// 参考: tbsrc/main/include/TradeBotUtils.h:237-504
type ThresholdSet struct {
	// 布尔标志
	UseNotional     bool // USE_NOTIONAL
	UsePercent      bool // USE_PERCENT
	UsePriceLimit   bool // USE_PRICE_LIMIT
	UseAheadPercent bool // USE_AHEAD_PERCENT
	UseCloseCross   bool // USE_CLOSE_CROSS
	UsePassiveThold bool // USE_PASSIVE_THOLD
	UseLinearThold  bool // USE_LINEAR_THOLD
	QuoteMaxQty     bool // QUOTE_MAX_QTY
	ClosePNL        bool // CLOSE_PNL
	CheckPNL        bool // CHECK_PNL
	NewsFlat        bool // NEWS_FLAT

	// 入场/出场阈值
	BeginPlace  float64 // BEGIN_PLACE
	BeginRemove float64 // BEGIN_REMOVE
	LongPlace   float64 // LONG_PLACE
	LongRemove  float64 // LONG_REMOVE
	ShortPlace  float64 // SHORT_PLACE
	ShortRemove float64 // SHORT_REMOVE
	LongInc     float64 // LONG_INC

	// 高波动变体
	BeginPlaceHigh float64 // BEGIN_PLACE_HIGH
	LongPlaceHigh  float64 // LONG_PLACE_HIGH

	// 仓位大小
	Size          int32 // SIZE
	TASize        int32 // TA_SIZE
	BeginSize     int32 // BEGIN_SIZE
	MaxSize       int32 // MAX_SIZE
	PercentSize   int32 // PERCENT_SIZE
	PercentLevel  int32 // PERCENT_LEVEL
	NotionalSize  int32 // NOTIONAL_SIZE
	NotionalMaxSz int32 // NOTIONAL_MAX_SIZE
	SMSRatio      int32 // SMS_RATIO
	MaxOSOrder    int32 // MAX_OS_ORDER
	BidSize       int32 // BID_SIZE
	BidMaxSize    int32 // BID_MAX_SIZE
	AskSize       int32 // ASK_SIZE
	AskMaxSize    int32 // ASK_MAX_SIZE

	// 激进单
	Cross        float64 // CROSS
	CloseCross   float64 // CLOSE_CROSS
	CloseImprove float64 // CLOSE_IMPROVE
	Improve      float64 // IMPROVE
	MaxCross     int32   // MAX_CROSS
	MaxLongCross int32   // MAX_LONG_CROSS
	MaxShortCross int32  // MAX_SHORT_CROSS
	CrossTarget  int32   // CROSS_TARGET
	CrossTicks   int32   // CROSS_TICKS
	AggCoolOff   int64   // AGG_COOL_OFF
	PlaceSpread  float64 // PLACE_SPREAD
	PILFactor    float64 // PIL_FACTOR

	// 风控
	StopLoss float64 // STOP_LOSS
	MaxLoss  float64 // MAX_LOSS
	UPNLLoss float64 // UPNL_LOSS
	PTProft  float64 // PT_PROFIT
	PTLoss   float64 // PT_LOSS
	MaxPrice float64 // MAX_PRICE
	MinPrice float64 // MIN_PRICE

	// 队列与大小
	OppQty        float64 // OPP_QTY
	SuppTolerance int     // SUPP_TOLERANCE
	AheadPercent  float64 // AHEAD_PERCENT
	AheadSize     float64 // AHEAD_SIZE
	SzAheadNoCxl  int     // SZAHEAD_NOCXL
	BookSzNoCxl   int     // BOOKSZ_NOCXL
	AggFlatBookSz int     // AGGFLAT_BOOKSIZE
	AggFlatBookFr float64 // AGGFLAT_BOOKFRAC

	// 套利
	Alpha        float64 // ALPHA
	SpreadEWA    float64 // SPREAD_EWA
	AvgSpreadAway int    // AVG_SPREAD_AWAY
	HedgeRatio   float64 // HEDGE_RATIO
	HedgeThres   float64 // HEDGE_THRES
	HedgeSzRatio float64 // HEDGE_SIZE_RATIO
	Const        float64 // CONST
	PriceRatio   float64 // PRICE_RATIO

	// Slop
	Slop int // SLOP

	// 时间
	Pause         int64 // PAUSE
	CancelReqPause int64 // CANCELREQ_PAUSE
	SqrOffTime    int64 // SQROFF_TIME
	SqrOffAgg     int   // SQROFF_AGG

	// Quote 相关
	QuoteSkew     float64 // QUOTE_SKEW
	MaxQuoteSpread int32  // MAX_QUOTE_SPREAD
	MaxQuoteLevel int     // MAX_QUOTE_LEVEL
	QuoteSignal   int     // QUOTE_SIGNAL

	// 统计
	StatDurationSmall int64   // STAT_DURATION_SMALL
	StatDurationLong  int64   // STAT_DURATION_LONG
	StatTradeThresh   float64 // STAT_TRADE_THRESH
	StatDecay         int     // STAT_DECAY

	// Delta
	DeltaHedge    float64 // DELTA_HEDGE
	TargetDelta   float64 // TARGET_DELTA
	MaxDeltaValue float64 // MAX_DELTA_VALUE
	MinDeltaValue float64 // MIN_DELTA_VALUE
	MaxDeltaChange float64 // MAX_DELTA_CHANGE

	// VWAP
	VWAPRatio float64 // VWAP_RATIO
	VWAPCount float64 // VWAP_COUNT
	VWAPDepth float64 // VWAP_DEPTH
	BidAskRatio float64 // BIDASK_RATIO

	// 支撑/拖尾
	SupportingOrders int32 // SUPPORTING_ORDERS
	TailingOrders    int32 // TAILING_ORDERS
	MaxOrders        int32 // MAX_ORDERS

	// PCA
	PCACoeff1 float64 // PCA_COEFF1
	PCACoeff2 float64 // PCA_COEFF2
	PCACoeff3 float64 // PCA_COEFF3

	// 杂项
	MinExtrInd   int     // MIN_EXTR_IND
	TargetStdDev float64 // TARGET_STD_DEV
	PriceCooloff int     // PRICE_COOLOFF

	// Sweep
	SweepPlace      int32 // SWEEP_PLACE
	SweepClose      int32 // SWEEP_CLOSE
	SweepPlaceLevel int32 // SWEEP_PLACE_LEVEL
	SweepCloseLevel int32 // SWEEP_CLOSE_LEVEL

	// tvar/tcache
	TVarKey   int32 // TVAR_KEY
	TCacheKey int32 // TCACHE_KEY
}

// NewThresholdSet 创建带 C++ 默认值的 ThresholdSet
// 参考: tbsrc/main/include/TradeBotUtils.h 构造函数 (lines 237-320)
func NewThresholdSet() *ThresholdSet {
	return &ThresholdSet{
		// 布尔默认值
		UseNotional:     false,
		UsePercent:      false,
		UsePriceLimit:   false,
		UseAheadPercent: false,
		UseCloseCross:   false,
		UsePassiveThold: false,
		UseLinearThold:  false,
		QuoteMaxQty:     false,
		ClosePNL:        true,
		CheckPNL:        true,
		NewsFlat:        false,

		// 仓位
		MaxOSOrder:   5,
		PercentLevel: 1,

		// 风控 — 极大值表示"不启用"
		UPNLLoss: 10000000000,
		StopLoss: 10000000000,
		MaxLoss:  100000000000,
		PTProft:  1000000,
		PTLoss:   1000000,
		MaxPrice: 1000000000000,
		MinPrice: -1000,

		// 队列
		OppQty:        1000000000,
		SuppTolerance: 1,
		AheadPercent:  100,
		AheadSize:     1000000000000,
		SzAheadNoCxl:  1000000,
		BookSzNoCxl:   1000000,

		// 激进单 — 极大值表示"不启用"
		Cross:         1000000000,
		CloseCross:    100000000000,
		Improve:       1000000000,
		MaxCross:      1000000000,
		MaxLongCross:  1000000000,
		MaxShortCross: 1000000000,
		CloseImprove:  -1,

		// 套利
		SpreadEWA:     0.6,
		AvgSpreadAway: 20,
		Slop:          20,

		// Quote
		MaxQuoteSpread: 1000000000,
		MaxQuoteLevel:  3,

		// VWAP
		VWAPRatio:   1,
		VWAPCount:   100,
		VWAPDepth:   10,
		BidAskRatio: 1,

		// 统计
		StatDurationLong: 1,
		StatDecay:        5,

		// Delta
		DeltaHedge:     100000,
		MaxDeltaValue:  1,
		MinDeltaValue:  -1,
		MaxDeltaChange: 2,

		// Sweep
		SweepPlaceLevel: 0,
		SweepCloseLevel: 0,

		// tvar/tcache
		TVarKey:   -1,
		TCacheKey: -1,

		// BID/ASK 独立大小
		BidSize:    0,
		BidMaxSize: 0,
		AskSize:    0,
		AskMaxSize: 0,
	}
}

File: tbsrc-java/src/main/java/com/quantlink/trader/core/ThresholdSet.java
package com.quantlink.trader.core;

/**
 * 策略阈值参数集。
 * 迁移自: tbsrc/main/include/TradeBotUtils.h — struct ThresholdSet (line 237-504)
 *
 * 所有默认值与 C++ ThresholdSet() 构造函数完全一致。
 */
public class ThresholdSet {

    // ---- 开关参数 ----
    // 迁移自: TradeBotUtils.h:241-251
    public boolean USE_NOTIONAL = false;
    public boolean USE_PERCENT = false;
    public boolean USE_PRICE_LIMIT = false;
    public boolean USE_AHEAD_PERCENT = false;
    public boolean USE_CLOSE_CROSS = false;
    public boolean USE_PASSIVE_THOLD = false;
    public boolean USE_LINEAR_THOLD = false; // C++: int USE_LINEAR_THOLD = 0
    public boolean QUOTE_MAX_QTY = false;
    public boolean CLOSE_PNL = true;
    public boolean CHECK_PNL = true;
    public boolean NEWS_FLAT = false;

    // ---- 仓位参数 ----
    // 迁移自: TradeBotUtils.h:336-345
    public int SIZE;
    public int TA_SIZE;
    public int BEGIN_SIZE;
    public int MAX_SIZE;
    public int PERCENT_SIZE;
    public int PERCENT_LEVEL = 1;
    public int NOTIONAL_SIZE;
    public int NOTIONAL_MAX_SIZE;
    public int SMS_RATIO;

    // ---- 买卖方向参数 ----
    // 迁移自: TradeBotUtils.h:496-503
    public int BID_SIZE = 0;
    public int BID_MAX_SIZE = 0;
    public int ASK_SIZE = 0;
    public int ASK_MAX_SIZE = 0;

    // ---- 风控参数 ----
    // 迁移自: TradeBotUtils.h:252-260
    public double OPP_QTY = 1_000_000_000;
    public int SUPP_TOLERANCE = 1;
    public double AHEAD_PERCENT = 100;
    public double AHEAD_SIZE = 1_000_000_000_000.0;
    public int SZAHEAD_NOCXL = 1_000_000;
    public int BOOKSZ_NOCXL = 1_000_000;
    public int AGGFLAT_BOOKSIZE = 0;
    public double AGGFLAT_BOOKFRAC = 0;
    public int MAX_OS_ORDER = 5;

    // ---- PNL/止损参数 ----
    // 迁移自: TradeBotUtils.h:261-267
    public double UPNL_LOSS = 10_000_000_000.0;
    public double STOP_LOSS = 10_000_000_000.0;
    public double MAX_LOSS = 100_000_000_000.0;
    public double PT_LOSS = 1_000_000;
    public double PT_PROFIT = 1_000_000;
    public double LONG_INC = 0;
    public double MAX_PRICE = 1_000_000_000_000.0;
    public double MIN_PRICE = -1000;

    // ---- 阈值参数 ----
    // 迁移自: TradeBotUtils.h:365-370
    public double BEGIN_PLACE;
    public double BEGIN_REMOVE;
    public double LONG_PLACE;
    public double LONG_REMOVE;
    public double SHORT_PLACE;
    public double SHORT_REMOVE;

    // ---- StdDev 阈值 ----
    // 迁移自: TradeBotUtils.h:419-424
    public double BEGIN_PLACE_STDDEV;
    public double LONG_PLACE_STDDEV;
    public double BEGIN_REMOVE_STDDEV;
    public double LONG_REMOVE_STDDEV;
    public double SHORT_PLACE_STDDEV;
    public double SQR_OFF_STDEV;

    // ---- VWAP ----
    // 迁移自: TradeBotUtils.h:272-275
    public double VWAP_RATIO = 1;
    public double VWAP_COUNT = 100;
    public double VWAP_DEPTH = 10;
    public double BIDASK_RATIO = 1;

    // ---- EWA / Spread ----
    // 迁移自: TradeBotUtils.h:276-284
    public double SPREAD_EWA = 0.6;
    public double CLOSE_CROSS = 100_000_000_000.0;
    public double CROSS = 1_000_000_000;
    public int CROSS_TARGET = 0;
    public int CROSS_TICKS = 0;
    public double IMPROVE = 1_000_000_000;
    public long AGG_COOL_OFF = 0;
    public double PLACE_SPREAD = 0.0;
    public double PIL_FACTOR = 0.0;

    // ---- CROSS 限制 ----
    // 迁移自: TradeBotUtils.h:285-291
    public int MAX_CROSS = 1_000_000_000;
    public int MAX_LONG_CROSS = 1_000_000_000;
    public int MAX_SHORT_CROSS = 1_000_000_000;
    public int MAX_QUOTE_SPREAD = 1_000_000_000;
    public double CLOSE_IMPROVE = -1;
    public double QUOTE_SKEW = 0;
    public double DELTA_HEDGE = 100_000;

    // ---- 时间参数 ----
    // 迁移自: TradeBotUtils.h:292-294
    public long PAUSE = 0;
    public long CANCELREQ_PAUSE = 0;
    public int MAX_QUOTE_LEVEL = 3;

    // ---- Sweep ----
    // 迁移自: TradeBotUtils.h:346-350
    public int SWEEP_PLACE;
    public int SWEEP_CLOSE;
    public int SWEEP_PLACE_LEVEL = 0;
    public int SWEEP_CLOSE_LEVEL = 0;
    public int SUPPORTING_ORDERS = 0;
    public int MAX_ORDERS = 0;
    public int TAILING_ORDERS = 0;

    // ---- Delta ----
    // 迁移自: TradeBotUtils.h:358-360
    public int DELTA;
    public int MAX_DELTA;

    // ---- PCA 系数 ----
    // 迁移自: TradeBotUtils.h:361-363
    public double PCA_COEFF1 = 0;
    public double PCA_COEFF2 = 0;
    public double PCA_COEFF3 = 0;
    public int QUOTE_SIGNAL = 0;

    // ---- 平仓时间 ----
    // 迁移自: TradeBotUtils.h:305-307
    public long SQROFF_TIME = 0;
    public int SQROFF_AGG = 0;
    public double TARGET_DELTA = 0;

    // ---- 统计参数 ----
    // 迁移自: TradeBotUtils.h:308-316
    public long STAT_DURATION_SMALL = 0;
    public long STAT_DURATION_LONG = 1;
    public double STAT_TRADE_THRESH = 0;
    public int STAT_DECAY = 5;
    public double MAX_DELTA_VALUE = 1;
    public double MIN_DELTA_VALUE = -1;
    public double MAX_DELTA_CHANGE = 2;

    // ---- 追单参数（pqr 20240902） ----
    // 迁移自: TradeBotUtils.h:490-493
    public int AVG_SPREAD_AWAY = 20;
    public int SLOP = 20;

    // ---- SPD hedge ----
    // 迁移自: TradeBotUtils.h:494-495
    public double CONST = 0.0;

    // ---- 其他 ----
    // 迁移自: TradeBotUtils.h 其余参数
    public int MIN_EXTR_IND = 0;
    public double TARGET_STD_DEV = 0.0;
    public int PRICE_COOLOFF = 0;
    public double ALPHA;
    public double HEDGE_RATIO;
    public double HEDGE_THRES;
    public double HEDGE_SIZE_RATIO;
    public double PRICE_RATIO;
    public double BEGIN_PLACE_HIGH;
    public double LONG_PLACE_HIGH;
    public int TWO_SIDED_QUOTE;
    public double TICKSIZE;
    public double DECAY;
    public double DECAY1;
    public double DECAY2;
    public long WINDOW_DURATION;
    public long LOOKBACK_TIME;
    public double HISTORICAL_STDDEV;
    public double TRGT_STDDEV;
    public double VOLATILITY_CONST;
    public long VOLATILITY_TIME;
    public double ARCH_COEFF;
    public double GARCH_COEFF;
    public int DEBUG_STMTS;
    public double CLOSE_SPREAD;
    public double CLOSING_SPREAD_DEV;
    public long SQUARE_OFF_TIME;
    public long SKIP_TIME;
    public int MAX_IMPROVE;
    public int CROSS_FLAG;
    public int ENTRY_BASED_SIGNAL;
    public long MEAN_DURATION_WINDOW;
    public int LOCAL_STD_TYPE;
    public double LOCAL_DEVIATION_WEIGHTAGE;
    public double BASE_DEVIATION_WEIGHTAGE;
    public int STDCOMP_VER;
    public double STDDEV_LP;
    public double STDDEV_LR;
    public double STDDEV_BP;
    public double STDDEV_BR;
    public double STDDEV_SP;
    public double STDDEV_SR;
    public double ERR_STDEV;
    public double LEARNING_RATE;
    public double STDEV_IMPROVE;
    public double STDEV_CROSS;
    public int SPREAD_COVER;
    public int IMMEDIATE_POS_CLOSE;

    // 迁移自: TradeBotUtils.h:500-503
    public int TVAR_KEY = -1;
    public int TCACHE_KEY = -1;
    public double UNDERLYING_UPPER_BOND = -1;
    public double UNDERLYING_LOWER_BOND = -1;

    // 迁移自: TradeBotUtils.h:474 — product_name
    public String productName = "";

    // ---- 以下字段来自 C++ ThresholdSet 完整定义 ----
    // 迁移自: TradeBotUtils.h:430-460

    // 模型/算法参数
    public int MODE_INSTRUMENT1;
    public int MODE_INSTRUMENT2;
    public int TARGET_PNL_MODE;
    public int StdevSqrOff_FLAG;
    public int LEADLAG_FLAG;
    public int DEP_BASKET;
    public int MODEL_ALGO;
    public int INP_FEAT_LENGTH;
    public int HIDDEN_NEURONS;
    public int LAGS;
    public int DYNAMIC_WEIGHTS;
    public double CONFIDENCE_INTERVAL_BEGIN;
    public double CONFIDENCE_INTERVAL_CLOSE;
    public int CONTINUOUS_TARGET_COMPUTATION;
    public int DYNAMIC_DEVIATION_COMPUTATION;

    // 模型文件路径
    // 迁移自: TradeBotUtils.h:467-472
    public String theta1_file = "";
    public String theta2_file = "";
    public String min_max_file = "";
    public String ar_bask0_file = "";
    public String ar_bask1_file = "";
    public String cov_mat_file = "";
}
